src/cluster/src/check/mod.rs

```diff
@@ -379,16 +398,208 @@ impl ClusterChecker {
+    pub async fn run_and_fix<F, R>(&self, fix: F) -> CheckResults
+    where
+        F: Fn(RecoverableCheck) -> R,
+        R: Future<Output = Result<(), UnrecoverableCheck>>,
+    {
+        // We want to collect all of the results of the checks
+        let mut results: Vec<CheckResult> = vec![];
+
+        for check in &self.checks {
+            // Perform one individual check
             let check_result = check.perform_check().await;
-            results.push(check_result);
+            match check_result {
+                // If the check passed, add it to the results list
+                it @ Ok(CheckStatus::Pass(_)) => results.push(it),
+                // If the check failed but is potentially auto-recoverable, try to recover it
+                Ok(CheckStatus::Fail(CheckFailed::AutoRecoverable(it))) => {
+                    let err = format!("{}", it);
+                    let fix_result = fix(it).await;
+                    match fix_result {
+                        // If the fix worked, return a passed check
+                        Ok(_) => results.push(Ok(CheckStatus::pass(format!("Fixed: {}", err)))),
+                        Err(e) => {
+                            // If the fix failed, wrap the original failed check in Unrecoverable
+                            results.push(Ok(CheckStatus::fail(CheckFailed::Unrecoverable(e))));
+                            // We return upon the first check failure
+                            return results;
+                        }
+                    }
+                }
+                it @ Ok(CheckStatus::Fail(_)) => {
+                    results.push(it);
+                    return results;
+                }
+                it @ Err(_) => {
+                    results.push(it);
+                    return results;
+                }
+            }
         }
 
-        CheckResults::from(results)
+        results
+    }
+
+    /// Performs all checks in an async task, returning the results via a channel.
+    ///
+    /// This function will return immediately with a channel which will yield progress
+    /// updates about checks as they are run.
+    ///
+    /// If you want to run the checks as a single batch and receive all of the results
+    /// at once, use [`run`] instead.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use fluvio_cluster::ClusterChecker;
+    /// # async fn do_run_with_progress() {
+    /// use async_channel::Receiver;
+    /// let progress: Receiver<CheckResult> = ClusterChecker::empty()
+    ///     .with_preflight_checks()
+    ///     .run_with_progress();
+    /// while let Ok(check_result) = progress.recv().await {
+    ///     println!("Got check result: {:?}", check_result);
+    /// }
+    /// # }
+    /// ```
+    pub fn run_with_progress(self) -> Receiver<CheckResult> {
+        let (sender, receiver) = async_channel::bounded(100);
+        spawn(async move {
```

Comment:
any reason why this need to be spawn?

Reply:
Yes, this has to do with binary/library separation. `run_with_progress` now returns immediately so that the CLI does not need to wait for all checks to complete before regaining control. The checks are performed inside the spawned task, and as they complete, they are sent to the caller via the Receiver that was given.

Reply:
How does this communicate end of progress?